### PR TITLE
TcpTransport: Use a preallocated buffer.

### PR DESCRIPTION
**Pull request by pspoerri**

### CODE REVIEW

**Review by BernardMetzler — COMMENTED** [review]
hard to parse. why class TcpPeer had to move around.

I would add the scratch buffer just to TcpSendState. no memory leak, no
fancy push/pop. also gives better locality that buffer can immediately be reused after tcp send/write call, since copied by socket layer.

**Comment by pspoerri** [comment]
> hard to parse. why class TcpPeer had to move around.
Fixed.

> I would add the scratch buffer just to TcpSendState. no memory leak, no fancy push/pop. also gives better locality. That buffer can immediately be reused after tcp send/write call, since copied by socket layer.
The TcpEndpoint/TcpSendState is configured after the buffer was read.
